rewritten items&tags functions to redux actions

brtsmrtn opened the pull request:
I have rewritten current functionalities to redux actions.

I have uncertainties concerning:
1. Implementation of action dispatchers in [items.ts](https://github.com/brtsmrtn/lb_2/pull/40/files#diff-de74ab54cb89b318f0d64ef584daf62a3ae1f479ff91a8c53f0605ea26c1d28d). I have reused some most of the functions as they were and mixed it up with more redux-like patterns in ASSIGN_TAG_TO_ITEM. I am aware of the fact that it is a mess and I am keeping it here as a stepping stone for other refactoring.
2. Action creators in [items.ts](https://github.com/brtsmrtn/lb_2/pull/40/files#diff-de74ab54cb89b318f0d64ef584daf62a3ae1f479ff91a8c53f0605ea26c1d28d). Is it alright that all of them have different `payload`? 
3. Batching/chaining dispatchers. Action `updateKnownTags` in [TagButton component](https://github.com/brtsmrtn/lb_2/pull/40/files#diff-b83d534aa6576fdfcbed9cbb386ca52c4ce80c7f91e389e50537627a387bd325) doesn't seem to update state directly, meaning that after dispatch I can't (seem to) get updated state of `knownTags` and immediately append this tag to the item. This is recognizable of the app behavior and user must click once to add an unknown tag and click for the second time to assign this tag (already known) to the item.

I would be glad if we could have a look at these points tomorrow @erikvanek , thanks ahead!

brtsmrtn committed:
rewritten items&tags functions to redux actions

brtsmrtn commented on src/components/TagButton.tsx:
```diff
@@ -64,23 +62,24 @@ export const TagButton: (props: TagButtonProps) => JSX.Element = ({
-          } else {
-            setErrorTag(errorTags.wrong);
-          }
+          dispatch(updateKnownTags(tagTitle));
```
Right after this action, another action should be dispatched:
`dispatch(assignTagToItem(X, item))`
... where X stands for previously created tag, which I can't seem to get neither from filtering it from the knownTags, nor reusing it from the return of the previous action.

Would you @erikvanek help a no-bro out?

erikvanek replied:
how about doing something like `const newTag = updateKnownTags(tagTitle)` and then `dispatch(newTag); dispatch(assignTagToItem(newTag.payload.tag, item))`? Not sure if it works though.

erikvanek commented on src/components/ListingItem.tsx:
```diff
@@ -1,50 +1,38 @@
+          color="primary"
+          onClick={() => dispatch(changeItemStatus(item))}
+        >
+          {!!item.alreadyRead ? "Unread" : "Read"}
```
not relevant for this PR but I think the double bang is not needed here. Even if the `.alradyRead` property is undefined, it should still render "Read"

erikvanek commented on src/components/ListingItem.tsx:
```diff
@@ -1,50 +1,38 @@
+        <Button
+          variant="outlined"
+          color="primary"
+          onClick={() => dispatch(changeItemStatus(item))}
```
Maybe I'd name the action creator a bit more specific, something like `markItemAsRead(item)`. `changeItemStatus(...)` feels not as self-explanatory as it could be imo.

erikvanek commented on src/components/TagButton.tsx:
```diff
@@ -64,23 +62,24 @@ export const TagButton: (props: TagButtonProps) => JSX.Element = ({
-    if (e.key === "Enter" && targetTagTitle) {
-      const tagAlreadyAssigned = isTagNew(targetTagTitle, item.tags);
-      const tagAlreadyKnown = isTagNew(targetTagTitle, knownTags);
+    const tag = e.target as HTMLTextAreaElement;
```
Changes in this codeblock are just so nice!

erikvanek commented on src/components/TagButton.tsx:
```diff
@@ -147,9 +146,7 @@ export const TagButton: (props: TagButtonProps) => JSX.Element = ({
-            renderTags={() => (
-              <TagList editable={true} item={item} tagDeleted={tagDeleted} />
-            )}
+            renderTags={() => <TagList editable={true} item={item} />}
```
Why it has to be done like this (anonymous function)? Why not just pure markup?

erikvanek commented on src/features/items.ts:
```diff
@@ -0,0 +1,134 @@
+  item: ListItem
+) => AssignTagToItemActionType = (tag, item) => {
+  return {
+    type: ASSIGN_TAG_TO_ITEM,
```
I think this logic should happen within appropriate part of reducer (in this case line 89 and below), not in action creator. While I think action creator can do things like `addNewItem()` does, in this case we're modifying incoming parameters and this does not feel good.

Imo a `AssignTagToItemActionType` should have three properties:
```
{
type: typeof ASSIGN_TAG_TO_ITEM;
tag: Tag;
item: ListItem
}
```
and reducer handles the logic. By doing that we're getting rid of genericly named `payload` property. It's completely fine if returned action has more properties than just the generic `type` & `payload`. Also, it's kinda usual to have a `type` property there, `payload` is not much self-explaining.

erikvanek commented on src/features/items.ts:
```diff
@@ -0,0 +1,134 @@
+export const UNASSIGN_TAG_FROM_ITEM = "UNASSIGN_TAG_FROM_ITEM";
+export type UnassignTagFromItemActionType = {
+  type: typeof UNASSIGN_TAG_FROM_ITEM;
+  payload: { tag: Tag; item: ListItem };
```
Again, would get rid of `payload` and rather have two properties `tag` & `item` defined in this type. No need for one more level in hierarchy.

erikvanek commented on src/features/items.ts:
```diff
@@ -0,0 +1,134 @@
+export const CHANGE_ITEM_STATUS = "CHANGE_ITEM_STATUS";
+export type ChangeItemStatusActionType = {
+  type: typeof CHANGE_ITEM_STATUS;
+  payload: ListItem;
```
again would rename payload -> item / listItem

erikvanek commented on src/features/items.ts:
```diff
@@ -0,0 +1,134 @@
+];
+
+export type ItemsActions =
+  | AssignTagToItemActionType
```
Maybe I led you in wrong direction but I think all those actions should not have the `Type` suffix in their name. I.e. `AssignTagToItemActionType` should become `AssignTagToItemAction` etc.

erikvanek commented on src/features/items.ts:
```diff
@@ -0,0 +1,134 @@
+      return [...state, action.payload];
+    }
+    case ASSIGN_TAG_TO_ITEM: {
+      return state.map((item) => {
```
I think we should destructure the `state` property so that we don't modify the original value.

erikvanek commented on src/features/items.ts:
```diff
@@ -0,0 +1,134 @@
+        return item;
+      });
+    }
+    case UNASSIGN_TAG_FROM_ITEM: {
```
this case would be happier with a test :)))

erikvanek commented on src/features/items.ts:
```diff
@@ -0,0 +1,134 @@
+      );
+      return newItems;
+    }
+    case CHANGE_ITEM_STATUS: {
```
When reading through the logic, how about having `TOGGLE_ITEM_STATUS` instead of `CHANGE_ITEM_STATUS`? To me toggle sounds more like what we're doing here.

erikvanek commented:
First of all gread job! Apart from some minor things I pointed out, much much better than the first attempt!


> 1. Implementation of action dispatchers in [items.ts](https://github.com/brtsmrtn/lb_2/pull/40/files#diff-de74ab54cb89b318f0d64ef584daf62a3ae1f479ff91a8c53f0605ea26c1d28d). I have reused some most of the functions as they were and mixed it up with more redux-like patterns in ASSIGN_TAG_TO_ITEM. I am aware of the fact that it is a mess and I am keeping it here as a stepping stone for other refactoring.

I think it's good. Just as I wrote in one of those comments, imo logic should be in the reducer. Some parsing / inferring manipulatoins can be done within the action creator. But changing propety values / reassigning values at should be done inside a reducer.

> 2. Action creators in [items.ts](https://github.com/brtsmrtn/lb_2/pull/40/files#diff-de74ab54cb89b318f0d64ef584daf62a3ae1f479ff91a8c53f0605ea26c1d28d). Is it alright that all of them have different `payload`?

Again already some comments there.

My point on this is that each action has its type definition (that defines properties that describe that action) and its action creator. Since each action does a different thing (say create an item X assign tag to an item), it is expected that all of them have different payloads. They do different things :)

> 3. Batching/chaining dispatchers. Action `updateKnownTags` in [TagButton component](https://github.com/brtsmrtn/lb_2/pull/40/files#diff-b83d534aa6576fdfcbed9cbb386ca52c4ce80c7f91e389e50537627a387bd325) doesn't seem to update state directly, meaning that after dispatch I can't (seem to) get updated state of `knownTags` and immediately append this tag to the item. This is recognizable of the app behavior and user must click once to add an unknown tag and click for the second time to assign this tag (already known) to the item.
> 
> I would be glad if we could have a look at these points tomorrow @erikvanek , thanks ahead!

I think once you dispatch an action, reducer gets evaluated and if the app state changes, the app rerenders. I think the solution that I sketched in the comment near that code block might work. Not sure but for sure we'll figure it out.

brtsmrtn committed:
resolved issues from review and added more tests

brtsmrtn commented:
Besides resolving issues/picking up on your hints raised in your review (thanks!) I have added few more tests and more importantly, removed the urlState to its own duck. I will be glad to know if you @erikvanek find some area for improvement (in terms of implementing Redux ofc) :)

erikvanek commented on package.json:
```diff
@@ -6,23 +6,27 @@
     "build": "react-scripts build && rd /s /q docs && move build docs",
     "gh-pages-deploy": "react-scripts build && rm -rf docs && mv build docs",
-    "test:watch": "jest --watch",
+    "test:watch": "jest --watch --no-coverage",
```
perfect :)

erikvanek commented on src/components/TagButton.tsx:
```diff
@@ -11,22 +14,19 @@ import {
+  const dispatch = useDispatch();
+  const knownTags = useSelector(
+    (state: ApplicationState) => state.knownTags
+  ).filter((t) => t.id !== "");
```
In which case a knownTag does not have an id?

erikvanek commented on src/features/items.ts:
```diff
@@ -0,0 +1,151 @@
+      return [...state, action.listItem];
+    }
+    case ASSIGN_TAG_TO_ITEM:
+      const indexOfItem = state.findIndex((item) => item.id === action.item.id);
```
In general what can be helpful (again to optimize the amount of characters needed to be typed) would be to extract the `item` property from `action` e.g. by doing `const {item} = action` call at the beginning of this code bloc. Cosmetic change, nothing big.

erikvanek commented on src/features/items.ts:
```diff
@@ -0,0 +1,151 @@
+    }
+    case ASSIGN_TAG_TO_ITEM:
+      const indexOfItem = state.findIndex((item) => item.id === action.item.id);
+      if (
```
I don't understand this. Does this mean that if we want to assign a same tag twice, nothing happens? If so, I'd try to prevent that action to happen. It's OK if reducer handles that as well, but I think the action should not be dispatched at all in that case.

erikvanek commented on src/features/items.ts:
```diff
@@ -0,0 +1,151 @@
+      }
+
+    case UNASSIGN_TAG_FROM_ITEM: {
+      if (
```
Same as above, is there a need for this if? Usually reducers don't have much if/else logic, that's why I'm a but curious here.

erikvanek commented on src/features/items.ts:
```diff
@@ -0,0 +1,151 @@
+    }
+    case TOGGLE_ITEM_STATUS: {
+      return [
+        Object.assign({}, action.item, {
```
I'd stick to use spread operator instead of Object.assign cause it's just shorter..

erikvanek reviewed (approved):
Awesome!

brtsmrtn committed:
resolved issues from review #2

brtsmrtn committed:
unifying actions in reducer
